### PR TITLE
Tests for erc20RootVault.sol

### PR DESCRIPTION
**Pull request by armoking32**
TODO:
increase coverage up to 100.

### CODE REVIEW

File: test/ERC20RootVault.test.ts [hunk 1]
```diff
@@ -0,0 +1,1149 @@
...
+                        .deposit(
+                            [BigNumber.from(10001), BigNumber.from(10001)],
+                            BigNumber.from(1),
+                            []
+                        )
+                ).to.emit(this.subject, "Deposit");
+            });
+
+            describe("edge cases:", () => {
+                describe("when deposit is enabled", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:688]: when deposit is disabled

File: test/ERC20RootVault.test.ts [hunk 2]
```diff
@@ -0,0 +1,1149 @@
...
+                            performanceFee: BigNumber.from(randomInt(10 ** 6)),
+                            depositCallbackAddress:
+                                ethers.constants.AddressZero,
+                            withdrawCallbackAddress:
+                                ethers.constants.AddressZero,
+                        };
+                        await this.erc20RootVaultGovernance
+                            .connect(this.admin)
+                            .stageDelayedStrategyParams(
+                                BigNumber.from(9),
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:733]: Magic constant?

File: test/ERC20RootVault.test.ts [hunk 3]
```diff
@@ -0,0 +1,1149 @@
...
+                            managementFee: BigNumber.from(randomInt(10 ** 6)),
+                            performanceFee: BigNumber.from(randomInt(10 ** 6)),
+                            depositCallbackAddress:
+                                ethers.constants.AddressZero,
+                            withdrawCallbackAddress:
+                                ethers.constants.AddressZero,
+                        };
+                        await this.erc20RootVaultGovernance
+                            .connect(this.admin)
+                            .stageDelayedStrategyParams(
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:732]: maybe I am wrong, but should this call be followed by commitDelayedStrategyParams after waiting for governanceDelay to pass
Otherwise delayedStrategyParams will not change

File: test/ERC20RootVault.test.ts [hunk 4]
```diff
@@ -0,0 +1,1149 @@
...
+                            this.subject.deposit(
+                                [BigNumber.from(10001), BigNumber.from(10001)],
+                                BigNumber.from(1),
+                                []
+                            )
+                        ).to.be.revertedWith(Exceptions.FORBIDDEN);
+                    });
+                });
+
+                describe("when minLpTokens more than lpAmount", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:746]: when minLpTokens is greater than lpAmount

File: test/ERC20RootVault.test.ts [hunk 5]
```diff
@@ -0,0 +1,1149 @@
...
+                        await this.usdc
+                            .connect(this.deployer)
+                            .approve(
+                                this.subject.address,
+                                BigNumber.from(10001)
+                            );
+                        await expect(
+                            this.subject
+                                .connect(this.deployer)
+                                .deposit(
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:817]: why do we receive 0 lpTokens amount in the second deposit? We should have received 10001 lpTokens if I am right, because we make a non-zero deposit => it is strange that we get 0 lpTokens

File: test/ERC20RootVault.test.ts [hunk 6]
```diff
@@ -0,0 +1,1149 @@
...
+                            .approve(this.subject.address, value);
+                        await expect(
+                            this.subject
+                                .connect(this.deployer)
+                                .deposit([value, value], BigNumber.from(1), [])
+                        ).to.be.revertedWith(Exceptions.LIMIT_OVERFLOW);
+                    });
+                });
+
+                describe("when sum of lpAmount and totalSupply is more than tokenLimit", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:858]: is greater than

File: test/ERC20RootVault.test.ts [hunk 7]
```diff
@@ -0,0 +1,1149 @@
...
+                                .connect(this.deployer)
+                                .deposit([value, value], BigNumber.from(1), [])
+                        ).to.be.revertedWith(Exceptions.LIMIT_OVERFLOW);
+                    });
+                });
+            });
+
+            describe("access control:", () => {
+                it("allowed: any address", async () => {
+                    await this.subject
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:890]: is it possible to move addDepositorsToAllowList into beforeEach to decrease code in each test?

File: test/ERC20RootVault.test.ts [hunk 8]
```diff
@@ -0,0 +1,1149 @@
...
+                        .connect(this.admin)
+                        .addDepositorsToAllowlist([this.deployer.address]);
+                    await withSigner(this.deployer.address, async (signer) => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:893]: access control: any address
However here you use only deployer as a signer, use randomAddress() helper instead

File: test/ERC20RootVault.test.ts [hunk 9]
```diff
@@ -0,0 +1,1149 @@
...
+        });
+
+        describe("#withdraw", () => {
+            it("emits Withdraw event", async () => {
+                await this.subject
+                    .connect(this.admin)
+                    .addDepositorsToAllowlist([this.deployer.address]);
+                await this.weth
+                    .connect(this.deployer)
+                    .approve(this.subject.address, BigNumber.from(10001));
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:930]: You could also move approve into beforeEach section 
Just to make sure, that every test will pass, give an approve by Uint256Max

File: test/ERC20RootVault.test.ts [hunk 10]
```diff
@@ -0,0 +1,1149 @@
...
+                await this.usdc
+                    .connect(this.deployer)
+                    .approve(this.subject.address, BigNumber.from(10001));
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:933]: same thing here

File: test/ERC20RootVault.test.ts [hunk 11]
```diff
@@ -0,0 +1,1149 @@
...
+                        )
+                ).not.to.be.reverted;
+
+                var to = randomAddress();
+                await expect(
+                    this.subject.withdraw(
+                        to,
+                        BigNumber.from(0),
+                        [BigNumber.from(0), BigNumber.from(0)],
+                        ["0x00", "0x00"]
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:950]: What do these 0x00 options for vaults pull methods mean? If you do not want to give any options, just leave empty bytes array

Comment (grigorevPavel) [reply test/ERC20RootVault.test.ts:950]: aka [[], []]

File: test/ERC20RootVault.test.ts [hunk 12]
```diff
@@ -0,0 +1,1149 @@
...
+                                randomAddress(),
+                                BigNumber.from(randomInt(100)),
+                                [],
+                                []
+                            )
+                        ).to.be.revertedWith(Exceptions.VALUE_ZERO);
+                    });
+                });
+
+                describe("when length of vaultsOptions and length of _subvaultNfts are differ", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:969]: "are different" or just "differ"

File: test/ERC20RootVault.test.ts [hunk 13]
```diff
@@ -0,0 +1,1149 @@
...
+                            "MockLpCallback",
+                            {
+                                from: this.deployer.address,
+                                args: [],
+                                log: true,
+                                autoMine: true,
+                            }
+                        );
+
+                        const lpCallback: Contract | MockLpCallback =
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:1052]: I think you can leave MockLpCallBack type only

File: test/ERC20RootVault.test.ts [hunk 14]
```diff
@@ -0,0 +1,1149 @@
...
+                                ["0x00", "0x00"]
+                            )
+                        ).to.emit(lpCallback, "WithdrawCallbackCalled");
+                    });
+                });
+            });
+
+            describe("access control:", () => {
+                it("allowed: any address", async () => {
+                    // when all edge cases will be done
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:1105]: delete comment

File: test/ERC20RootVault.test.ts [hunk 15]
```diff
@@ -0,0 +1,1149 @@
...
+                    await withSigner(randomAddress(), async (signer) => {
+                        await this.subject
+                            .connect(this.admin)
+                            .addDepositorsToAllowlist([this.deployer.address]);
+                        await this.weth
+                            .connect(this.deployer)
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:1111]: connect(signer)
do not forget to mint tokens for the ranndomAddress

File: test/ERC20RootVault.test.ts [hunk 16]
```diff
@@ -0,0 +1,1149 @@
...
+                            .approve(
+                                this.subject.address,
+                                BigNumber.from(10001)
+                            );
+                        await this.usdc
+                            .connect(this.deployer)
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:1117]: connect(signer)

File: test/ERC20RootVault.test.ts [hunk 17]
```diff
@@ -0,0 +1,1149 @@
...
+                            .approve(
+                                this.subject.address,
+                                BigNumber.from(10001)
+                            );
+                        await expect(
+                            this.subject
+                                .connect(this.deployer)
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:1124]: connect(signer)

File: test/ERC20RootVault.test.ts [hunk 18]
```diff
@@ -0,0 +1,1149 @@
...
+                                        BigNumber.from(10001),
+                                    ],
+                                    BigNumber.from(0),
+                                    []
+                                )
+                        ).not.to.be.reverted;
+
+                        var to = randomAddress();
+                        await expect(
+                            this.subject.withdraw(
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:1137]: by default we connect to this.deployer, but we have to make sure that withdraw works for randomAddress(), so pls add connect(signer)

File: test/ERC20RootVault.test.ts [hunk 19]
```diff
@@ -0,0 +1,1149 @@
...
+                    return this.subject;
+                }
+            );
+        });
+
+        beforeEach(async () => {
+            await this.deploymentFixture();
+        });
+
+        describe("#depositorsAllowlist", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:194]: pls add access control: allowed any address

File: test/ERC20RootVault.test.ts [hunk 20]
```diff
@@ -0,0 +1,1149 @@
...
+            it("returns non zero length of depositorsAllowlist", async () => {
+                expect(
+                    (await this.subject.depositorsAllowlist()).length
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:197]: a little bit strange test. It is way better to set a depositorsAllowList and compare actual result with the expected one

File: test/ERC20RootVault.test.ts [hunk 21]
```diff
@@ -0,0 +1,1149 @@
...
+                    "0x4", // address of _nft
+                ]);
+                await ethers.provider.send("hardhat_setStorageAt", [
+                    this.subject.address,
+                    "0x4", // address of _nft
+                    "0x0000000000000000000000000000000000000000000000000000000000000000",
+                ]);
+            });
+
+            xit("initializes contract successfully", async () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:339]: why xit?

File: test/ERC20RootVault.test.ts [hunk 22]
```diff
@@ -0,0 +1,1149 @@
...
+                                            randomAddress(),
+                                            [startNft]
+                                        )
+                                ).to.be.revertedWith(Exceptions.FORBIDDEN);
+                            }
+                        );
+                    });
+                });
+
+                describe("when subvaultNft index is 0 (Somehow works)", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:448]: in our deployment it means that rootVault has itself as subVault, because we deploy rootVault firstly => it gets 0 nft.
So remove (somehow works) and replace it with the following explanation or smth pls

File: test/ERC20RootVault.test.ts [hunk 23]
```diff
@@ -0,0 +1,1149 @@
...
+                            (
+                                await this.vaultRegistry.vaultsCount()
+                            ).toNumber() - 2;
+                        await ethers.provider.send("hardhat_setStorageAt", [
+                            this.vaultRegistry.address,
+                            "0x5", // address of nft index
+                            "0x0000000000000000000000000000000000000000000000000000000000000000",
+                        ]);
+                        await withSigner(
+                            this.erc20RootVaultGovernance.address, // what to write here????
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:487]: what is this comment about?)

File: test/ERC20RootVault.test.ts [hunk 24]
```diff
@@ -0,0 +1,1149 @@
...
+                                            randomAddress(),
+                                            [startNft]
+                                        )
+                                ).to.be.revertedWith(Exceptions.DUPLICATE);
+                            }
+                        );
+                    });
+                });
+
+                describe("when subvaultNft index is 0", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:475]: I think not subVaultNft index, but subVault index perhaps

File: test/ERC20RootVault.test.ts [hunk 25]
```diff
@@ -0,0 +1,1149 @@
...
+                                            randomAddress(),
+                                            [startNft]
+                                        )
+                                ).to.be.revertedWith(Exceptions.DUPLICATE);
+                            }
+                        );
+                    });
+                });
+
+                describe("when subvaultNft address is 0", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:507]: again subVault address

File: test/ERC20RootVault.test.ts [hunk 26]
```diff
@@ -0,0 +1,1149 @@
...
+                                            this.nft,
+                                            [
+                                                this.usdc.address,
+                                                this.weth.address,
+                                            ],
+                                            randomAddress(),
+                                            [startNft]
+                                        )
+                                ).to.be.revertedWith(
+                                    Exceptions.INVALID_INTERFACE // ADDRESS_ZERO
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:534]: Does not match with description above, rewrite the test pls. It is possible that description is wrong

File: test/ERC20RootVault.test.ts [hunk 27]
```diff
@@ -0,0 +1,1149 @@
...
+                                        )
+                                ).to.be.revertedWith(
+                                    Exceptions.INVALID_INTERFACE
+                                );
+                            }
+                        );
+                    });
+                });
+
+                describe("when subvaultNFT supports interface", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:567]: WTF )))
even if you mean subVault instead of subVaultNft, if a subVault does NOT support the required interface, the transaction must be reverted with INVALID_INTERFACE

File: test/ERC20RootVault.test.ts [hunk 28]
```diff
@@ -0,0 +1,1149 @@
...
+                    it(`reverts with ${Exceptions.INVALID_INTERFACE}`, async () => {
+                        const startNft = (
+                            await this.vaultRegistry.vaultsCount()
+                        ).toNumber();
+                        await withSigner(
+                            this.erc20VaultGovernance.address, // what to write here????
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:573]: again panic in comments ... oh my

File: test/ERC20RootVault.test.ts [hunk 29]
```diff
@@ -0,0 +1,1149 @@
...
+                                        )
+                                ).to.be.revertedWith(
+                                    Exceptions.INVALID_INTERFACE
+                                );
+                            }
+                        );
+                    });
+                });
+
+                // describe("when vault's nft is not 0", () => {
```
Comment (grigorevPavel) [inline test/ERC20RootVault.test.ts:595]: delete all these comments

**Review by alexeuler — APPROVED** [review]
